### PR TITLE
[SPARK-26577][SQL] Add input optimizer when reading Hive table by SparkSQL

### PR DESCRIPTION
**Pull request by Deegue**
## What changes were proposed in this pull request?

When using SparkSQL, for example the ThriftServer, if we set

`spark.sql.hive.fileInputFormat.enabled=true`

we can optimize the InputFormat to CombineTextInputFormat automatically if it's TextInputFormat before. And we can also change the max/min size of input splits by setting, for example 

`spark.sql.hive.fileInputFormat.split.maxsize=268435456`
`spark.sql.hive.fileInputFormat.split.minsize=134217728`

Otherwise, we have to modify Hive Configs and structure of tables.

And we made a test by using a Hive table with a lot of small files in HDFS and haven't combined :

Before improved:
image

After improved:
image

## How was this patch tested?

Added a test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -561,6 +561,23 @@ object SQLConf {
     .checkValues(HiveCaseSensitiveInferenceMode.values.map(_.toString))
     .createWithDefault(HiveCaseSensitiveInferenceMode.INFER_AND_SAVE.toString)
 
+  val HIVE_FILE_INPUT_FORMAT_ENABLED = buildConf("spark.sql.hive.fileInputFormat.enabled")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: These are Hive only. Thus, please move these to `object HiveUtils`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 2]
```diff
@@ -123,8 +123,26 @@ class HadoopTableReader(
     val inputPathStr = applyFilterIfNeeded(tablePath, filterOpt)
 
     // logDebug("Table input: %s".format(tablePath))
-    val ifc = hiveTable.getInputFormatClass
+    var ifc = hiveTable.getInputFormatClass
       .asInstanceOf[java.lang.Class[InputFormat[Writable, Writable]]]
+    if (conf.fileInputFormatEnabled) {
+      hadoopConf.set("mapreduce.input.fileinputformat.split.maxsize",
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: dedup the code?

**Comment by gatorsmile** [comment]
Let me think about the conf naming. Could you first rebase it to master branch?  Now it is targeting to 2.4.

**Comment by Deegue** [comment]
> Let me think about the conf naming. Could you first rebase it to master branch? Now it is targeting to 2.4.

Thank you, I've picked the changes to master branch and create a new PR.
#23506

**Comment by gatorsmile** [comment]
Let us close this one and will review the PR in the master branch.

**Pull request closed without merging**